File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> output = new ArrayList<>();
        solve(0, candidates, target, ans, output);
        return ans;
    }
    
    public void solve(int index, int[] A, int target, List<List<Integer>> ans, List<Integer> output){
        if(target == 0){
            ans.add(new ArrayList<>(output));
            return;
        }
        if(index==A.length || target<0){
            return;
        }
        output.add(A[index]);
        solve(index,A,target-A[index], ans, output);
        output.remove(output.size()-1);
        solve(index+1, A, target, ans, output);
        
    }
}

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        def backtracking(path, remaining, start):
            if remaining == 0:
                ans.append(path[:])
                return
            if remaining < 0:
                return
            for i in range(start, len(candidates)):
                path.append(candidates[i])
                backtracking(path, remaining - candidates[i], i)
                path.pop()
        backtracking([], target, 0)
        return ans